tests/phpunit/CRM/MembershipExtras/Service/MembershipInstalmentsScheduleTest.php

```diff
@@ -0,0 +1,183 @@
+      2
+    );
+
+    $instalment = $instalments[0];
+    $this->assertEquals($expectedAmount, $instalment->getInstalmentAmount()->getAmount());
+  }
+
+  /**
+   * Test Generate Rolling Annually Instalment
+   *
+   * @throws Exception
+   */
+  public function testCalculateRollingAnnuallyInstalments() {
+    $membershipTypes = $this->getRollingMembershipTypes();
+    $instalments = $this->getRollingMembershipInstalments(
+      $membershipTypes,
+      MembershipInstalmentsSchedule::ANNUALLY
+    );
+
+    $this->assertCount(MembershipInstalmentsSchedule::ANNUALLY_INSTALMENT_COUNT, $instalments);
+
+    $expectedAmount = MoneyUtilities::roundToPrecision(
+      ($membershipTypes[0]->minimum_fee + $membershipTypes[1]->minimum_fee) / MembershipInstalmentsSchedule::ANNUALLY_INSTALMENT_COUNT,
+      2
+    );
+
+    $instalment = $instalments[0];
+    $this->assertEquals($expectedAmount, $instalment->getInstalmentAmount()->getAmount());
+  }
+
+  /**
+   * Tests Exception when providing incorrect duration unit
+   *
+   * @throws CRM_MembershipExtras_Exception_InvalidMembershipTypeInstalmentCalculator
+   */
+  public function testExceptionIsThrownIfMembershipTypeDurationUnitIsNotYearly() {
+    $membershipType1 = MembershipTypeFabricator::fabricate([
+      'duration_unit' => 'month',
+    ]);
+    $membershipType2 = MembershipTypeFabricator::fabricate([
+      'duration_unit' => 'year',
+    ]);
+
+    $membershipTypes = [$membershipType1, $membershipType2];
+    $this->expectException(InvalidMembershipTypeInstalmentCalculator::class);
+    $this->getMembershipTypeInstalmentAmount($membershipTypes, MembershipInstalmentsSchedule::ANNUALLY);
+  }
+
+  /**
+   * Tests exception when providing incorrect duration interval
+   *
+   * @throws CRM_MembershipExtras_Exception_InvalidMembershipTypeInstalmentCalculator
+   */
+  public function testExceptionIsThrownIfMembershipTypeDurationUnitIsNotOneYear() {
+    $membershipType1 = MembershipTypeFabricator::fabricate([
+      'duration_unit' => 'year',
+      'duration_interval' => 2,
+    ]);
+    $membershipType2 = MembershipTypeFabricator::fabricate([
+      'duration_unit' => 'year',
+      'duration_interval' => 1,
+    ]);
+
+    $membershipTypes = [$membershipType1, $membershipType2];
+    $this->expectException(InvalidMembershipTypeInstalmentCalculator::class);
+    $this->getMembershipTypeInstalmentAmount($membershipTypes, MembershipInstalmentsSchedule::ANNUALLY);
+  }
+
+  /**
+   * @param $membershipTypes
+   * @param $schedule
+   * @return mixed
+   * @throws Exception
+   */
+  private function getRollingMembershipInstalments($membershipTypes, $schedule) {
+    $membershipTypeInstalmentCalculator = $this->getMembershipTypeInstalmentAmount($membershipTypes, $schedule);
+
+    $startDate = new DateTime();
+    $joinDate = $startDate;
+    $endDate = new DateTime();
+    $endDate->add(new DateInterval('P1Y'));
```

Comment:
the end date should be after X amount of time (as defined by the membership type) minus 1 day

Reply:
@omarabuhussein  currently, we support monthly for 1 year  rolling membership type at this stage, I will implement / refactor when we are implementing, quarterly, and annually.

Reply:
but the end date of the membership is related to its duration and not if it paid annually or monthly or etc, and it always be `-1 day` from the (start date + membership duration)